android/src/main/kotlin/carnegietechnologies/gallery_saver/GallerySaver.kt

```diff
@@ -0,0 +1,147 @@
+package carnegietechnologies.gallery_saver
+
+import android.Manifest
+import android.app.Activity
+import android.content.pm.PackageManager
+import androidx.core.app.ActivityCompat
+import io.flutter.plugin.common.MethodCall
+import io.flutter.plugin.common.MethodChannel
+import io.flutter.plugin.common.PluginRegistry
+import kotlinx.coroutines.*
+
+/**
+ * Class holding implementation of saving images and videos
+ */
+class GallerySaver internal constructor(private val activity: Activity) : PluginRegistry.RequestPermissionsResultListener {
+
+    private var pendingResult: MethodChannel.Result? = null
+    private var methodCall: MethodCall? = null
+    private var isImage: Boolean = false
+
+    private val job = Job()
+    private val uiScope = CoroutineScope(Dispatchers.Main + job)
+
+    /**
+     * Saves image or video to device
+     *
+     * @param methodCall - method call
+     * @param result     - result to be set when saving operation finishes
+     * @param isImage    - tells if image or video should be saved
+     */
+    internal fun saveFile(methodCall: MethodCall, result: MethodChannel.Result, isImage: Boolean) {
+        if (!setPendingMethodCallAndResult(methodCall, result)) {
+            finishWithError()
+            return
+        }
+
+        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE)
```

Comment:
maybe asking for permission should be done by the underlying library
and we should just return with error here.